### ISSUE TITLE
Glue is being EXTRA gluey when used in Function

### ISSUE THREAD

**Issue by lrcall** [inkle/inky#166]
I made a bug report in INK.  Load the example below as it will explain the issues it in detail.

-> lonedebug
=== lonedebug ===
#CLASS:__ui__
A series of weird glitches related to spacing between a variable called within a function and the space between it and glue being destoryed entirely. It will even pull in a tag out of the greyed out \#tag view and draw it in a normal story line view...
Error 0,1,2 and 3 highlight the destruction of the spacing found between the end of a variable and a glue marker. 
Error 4 is a another test case
Error 5 is my current hack
+ (parsing) [Parsing]
    #CLASS:__ui__
    ++ (glueerror)[Glue Errors]
        +++ [Glue Error 0]
            Pulls tags into inky viewer
            ~function0("nothing")
            #CLASS:data
        +++ [Glue Error 1]
            Pulls tags into inky viewer
            ~function1("nothing")
            #CLASS:data
        +++ [Glue Error 2]
            Expected ({function0("nothing")}{" "}{functionA("data")})        
            Is (<>
            ~function2("nothing")
            ~functionA("data")
            <>)
        +++ [Glue Error 3]
            Expected ({function0("nothing")}{" "}{functionA("data")})        
            Is (<>
            ~function3("nothing")
            ~functionA("data")
            <>)
        +++ [Glue Error 4]
            In this case the first empty space between the variable \{_value\} and \<\> is maintaned if a string detectable element (nbsp in this case) is placed after it. Any string placed between 2 glue connectors will preserve a single previous space placed between 2 glues or a variable and glue.
            Expected ({function1("nothing")}{"  "}{functionA("data")})
            Is (<>
            ~function1("nothing")
            ~functionA("data")
            <>)
        +++ [Glue Error 5]
            This is the only hack I figured out to fix the spacing issue, by embedding a space inside of a string blaced between brackets. Also placing the functions side by side within brackets and one on top of the other using \~ delimeter offer the same exact output so currently the only consistent method is this one.
            Expected ({function1("nothing")}{functionB("data")})
            Is (<>
            ~function1("nothing")
            ~functionB("data")
            <>)
        --- -> parsing
        
        
=== function function0(_value) ===
do:{_value} <>
=== function function1(_value) ===
do:{_value}{" "}<>
=== function function2(_value) ===
do:{_value}&nbsp<>
=== function function3(_value) ===
do:{_value} <>\&nbsp<>
=== function functionA(_asset) ===
{"text"}
=== function functionB(_asset) ===
{" {"text"}"}

**Comment by joningold** [comment]
So... I’m finding it hard to separate what’s going on here because I can’t
tell exactly what you’re trying to achieve!

In general, function calls made using a ~ mark can produce output but their
glue behaviour is a bit unpredictable because they’re called not as content
but as logic. You might find it easier to format using {...} notation:

Go to a {function(“some text”)} and then continue

This should work as expected an preserve the glue either side of the
function call.

The tag issue is likely because tags are added onto lines of content but in
the cases shown, there’s no content there! Remember that a tag isn’t really
an “instruction” that sits on its own line; it’s supposed to be a mark up
on a piece of output. Again {..} notation can help:

{function(some content)} #tag

Should work because the tag is now attached to a line of content - content
which as it happens calls a function.

Tl;dr {function(..)} is a better way to call a function that’s going to
produce output if you want to control the layout closely. ~ function() is
really intended for “here’s a line that doesn’t do any any text”

HTH

**Comment by lrcall** [comment]
I simplified the case to the following script. I hope it explains the weird inconsistency.
Using a {} in a function regardless if it is a string or a variable cause the same behavior

/====================================
>Non-Agressive Glue
{"glue"} <>text
{"glue"} <>
text
/====================================
>Agressive Glue
{function1()}{functionA()}
~function1()
~functionA()
/====================================
=== function function1() ===
{"glue"} <>
=== function functionA() ===
text

**Comment by joningold** [comment]
Thanks, that's very helpful!

To simplify further, I think the problem is the difference in output
between these two:

{Text0()}.
{Text1()}.

=== function Text0() ===
Text0 <>

=== function Text1() ===
{"Text1"} <>



Note, for now, that if possible

{function1()} {functionA()}

is the best way to fix the problem; or alternatively;

~ function1()
<> <>
~ functionA()

-- in both cases, the thing calling the function decides where to put the
spaces; rather than the function itself.
